src/utils/dimensionality_operations.jl

```diff
@@ -295,6 +295,256 @@ function wavelet_unsqueeze(Y::AbstractArray{T,5}; type=WT.db1) where T
     return X
 end
 
+function HaarLift(x::AbstractArray{T,4},dim) where T
+  #Haar lifting
+
+  # Splitting
+  if dim==1
+    L = x[2:2:end, :, :, :]
+    H = x[1:2:end, :, :, :]
+  elseif dim==2
+    L = x[:,2:2:end, :, :]
+    H = x[:,1:2:end, :, :]
+  else
+    error("dimension for Haar lifting has to be 1 or 2 for a 4D input array")
+  end
+
+  # predict
+  H .= H .- L
+
+  #update
+  L .= L .+ H ./ T(2.0)
+
+  #normalize
+  H .= H ./sqrt(T(2.0))
+  L .= L.*sqrt(T(2.0))
+
+  return L,H
+end
+
+function invHaarLift(L::AbstractArray{T,4},H::AbstractArray{T,4},dim) where T
```

Comment:
Can we try to avoid these `Array{T, 4}, Array{T, 5}` split, it's really verbose and a pain to maintain with all the replicated code.